Add tutorials

spyridon97 opened the pull request:
This PR fixes several documentation warnings, and adds tutorials for the basic ADIOS concepts

This PR fixes https://github.com/ornladios/ADIOS2/issues/3793.

anagainaru commented on docs/user_guide/source/tutorials/timesteps.rst:
```diff
@@ -0,0 +1,203 @@
+
+.. _sec:tutorials_basics_timesteps:
+
+In the previous tutorial, we introduced the concept of operators, and briefly touched upon the concept of timesteps.tutorial
```
Is this a typo (the ".tutorial" at the end of this sentence)?

anagainaru reviewed (commented):
Two questions/comments
1. You moved a bunch of examples from bpWriteRead{name} to bpTimeWriteRead{name}. What is the significance of Time in the name? I would keep the name the way it was just because it was shorter and still intuitive (unless there is a reason for the Time addition).
2. You added a bunch of skeleton codes (which is great for the tutorial text you have). I would place them all in a separate folder instead of the hello 

I left a bunch of comments with a few typos, but other than this it looks really good. Good job !

spyridon97 commented:
> 1. You moved a bunch of examples from bpWriteRead{name} to bpTimeWriteRead{name}. What is the significance of Time in the name? I would keep the name the way it was just because it was shorter and still intuitive (unless there is a reason for the Time addition).

I did it because bpWriteRead is kinda very vague, and there are a bunch of examples like that. This particular examples show how to write and read time steps. And that's how i also use them.

> 2. You added a bunch of skeleton codes (which is great for the tutorial text you have). I would place them all in a separate folder instead of the hello

@vicentebolea and I decided that the best place to put them is next to the actual code so that we don't forget to update them, if needed.

eisenhauer commented:
> 
> So you would prefer to rename the examples to bpStepWriteRead*?
Definitely would be my preference.  "Time" doesn't appear in our APIs, but "Step" is prominent.  ("Timestep" is in some SST engine parameters, but we likely should change that at some point.)

> And is it okay if i stick to timesteps for the name of the particular tutorial?
I'd ask @pnorbert to weigh in on that.  I know about the de-emphasis on Timestep in favor of Step, but not as much about the why's.  Norbert has done the demo and tutorials enough to opine on this.

spyridon97 commented:
> > So you would prefer to rename the examples to bpStepWriteRead*?
> > Definitely would be my preference.  "Time" doesn't appear in our APIs, but "Step" is prominent.  ("Timestep" is in some SST engine parameters, but we likely should change that at some point.)
> 
> > And is it okay if i stick to timesteps for the name of the particular tutorial?
> > I'd ask @pnorbert to weigh in on that.  I know about the de-emphasis on Timestep in favor of Step, but not as much about the why's.  Norbert has done the demo and tutorials enough to opine on this.

I decided to remove the word from everywhere in this set of tutorials, I hope addresses all of your concerns helps.

eisenhauer commented on docs/user_guide/source/components/engine.rst:
```diff
@@ -323,7 +325,7 @@ PerformsPuts
 
 
 PerformsDataWrite
```
Not an issue with this update, but everywhere that says "Performs*" in headers here should be "Perform" (without the s).

spyridon97 replied:
i fixed that too.

vicentebolea commented:
The crusher failure can be resolve if you "update the branch", _IMPORTANT_ use rebase update

vicentebolea commented:
![Screenshot from 2023-11-01 12-35-21](https://github.com/ornladios/ADIOS2/assets/939798/dceef762-f66a-428c-b0e1-045f6e1ad2d1)

spyridon97 commented:
@vicentebolea the bpls results were added, can you check them and approve this PR?

vicentebolea reviewed (commented):
The changes with regards the bpls usage looks good. :+1: 

As for the the several changes in the documentation unrelated to the tutorials, they look good in a first glance, however, while it is great to improve things here and there, it is best to do it in a different PR. The reason for this is that opening many fronts will both slow the review process and confuse reviewers.

spyridon97 commented:
> The changes with regards the bpls usage looks good. 👍
> 
> As for the the several changes in the documentation unrelated to the tutorials, they look good in a first glance, however, while it is great to improve things here and there, it is best to do it in a different PR. The reason for this is that opening many fronts will both slow the review process and confuse reviewers.

I will keep that in mind for future PRs.

vicentebolea commented on docs/user_guide/source/advanced/query.rst:
```diff
@@ -21,14 +21,14 @@ to construct a query and evaluate using the engine.
 
-	// touched_blocks is a list of regions specified by (start, count),
-	// that contains data that satisfies the query file
+	      // touched_blocks is a list of regions specified by (start, count),
+	      // that contains data that satisfies the query file
```
```suggestion
	    // touched_blocks is a list of regions specified by (start, count),
	    // that contains data that satisfies the query file
```

spyridon97 replied:
Fixed

vicentebolea commented on docs/user_guide/source/components/engine.rst:
```diff
@@ -361,13 +363,13 @@ The following table summarizes the memory contracts required by ADIOS2 engines b
 | Get      | Data Memory | Contract                                      |
 +----------+-------------+-----------------------------------------------+
-|          | Pointer     | do not modify until PerformPuts/EndStep/Close |
+|          | Pointer     | do not modify until PerformGets/EndStep/Close |
```
Is this correct? Shouldnt it be Put?

spyridon97 replied:
This is correct. There is another table for PerformPut

vicentebolea commented on docs/user_guide/source/ecosystem/h5vol/vol.rst:
```diff
@@ -1,11 +1,12 @@
 .. note::
 
-The Virtual Object Layer (VOL) is a feature introduced in recent release of  HDF5 1.12 (https://hdf5.wiki/index.php/New_Features_in_HDF5_Release_1.12). 
+  The Virtual Object Layer (VOL) is a feature introduced in recent release of  HDF5 1.12 (https://hdf5.wiki/index.php/New_Features_in_HDF5_Release_1.12).
```
Why this space at the beginning?

spyridon97 replied:
Fixed

vicentebolea commented on docs/user_guide/source/advanced/ecp_hardware.rst:
```diff
@@ -33,27 +33,29 @@ your tests.
- # On another terminal launch multiple instances of the Reader test
- srun {PROJFLAGS} -N 2 /gpfs/alpine/proj-shared/csc331/vbolea/ADIOS2-build/bin/TestCommonRead SST mpi_dp_test
+  # Launch simple writer test instance
+  srun {PROJFLAGS }-N 1 /gpfs/alpine/proj-shared/csc331/vbolea/ADIOS2-build/bin/TestCommonWrite SST mpi_dp_test CPCommPattern=Min,MarshalMethod=BP5
```
```suggestion
  srun {PROJFLAGS } -N 1 /gpfs/alpine/proj-shared/csc331/vbolea/ADIOS2-build/bin/TestCommonWrite SST mpi_dp_test CPCommPattern=Min,MarshalMethod=BP5
```

Not a change in this PR but good to change :)

spyridon97 replied:
Fixed

vicentebolea reviewed (commented):
Looks good, couple of comments, let the CI pass before making change so we can see the crusher CI result

spyridon97 commented:
@vicentebolea I believe the changes you suggested could be part of documentation improvements PR. and leave this PR as is.

vicentebolea commented on examples/hello/CMakeLists.txt:
```diff
@@ -3,31 +3,34 @@
 endif()
-
 find_package(hip QUIET)
 if(ADIOS2_HAVE_Kokkos_HIP OR hip_FOUND)
-  add_subdirectory(bpWriteReadHip)
+  add_subdirectory(bpStepsWriteReadHip)
 endif()
```
> Not to be addressed here, Note for another PR. 

For consistency across the CMake code we should not use find_package outside of detectoptions if within the project(adios2)

vicentebolea reviewed (approved):
Upon CI passing

spyridon97 commented:
> Upon CI passing

There seems to be unrelated tests failing. Shall i merge?

vicentebolea commented:
Indeed, it is unrelated, ready to merge!
